src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+        }
+    }
+}
+
+impl<Upstream, C> Connection<Upstream, C>
+where
+    C: Clock + Send + Sync + 'static,
+{
+    /// Create a new connection with default configuration parameters.
+    pub fn new_with_time(upstream: Upstream, clock: C) -> Self {
+        Self::with_time_config(upstream, clock, Default::default())
+    }
+
+    /// Create a new connection with specified configuration parameters.
+    pub fn with_time_config(
+        upstream: Upstream,
+        clock: C,
+        config: Config,
+    ) -> Self {
+        Self {
+            upstream,
+            cache: Cache::new(config.max_cache_entries),
+            config,
+            clock,
+        }
+    }
+}
+
+impl<Upstream> Debug for Connection<Upstream> {
+    fn fmt(&self, _: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
+        todo!()
+    }
+}
+
+//------------ SendRequest ----------------------------------------------------
+
+impl<CR, Upstream, C> SendRequest<CR> for Connection<Upstream, C>
+where
+    CR: Clone + ComposeRequest + 'static,
+    Upstream: Clone + SendRequest<CR> + Send + Sync + 'static,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    fn send_request(
+        &self,
+        request_msg: CR,
+    ) -> Box<dyn GetResponse + Send + Sync> {
+        Box::new(Request::<CR, Upstream, C>::new(
+            request_msg,
+            self.upstream.clone(),
+            self.cache.clone(),
+            self.config.clone(),
+            self.clock.clone(),
+        ))
+    }
+}
+
+//------------ Request --------------------------------------------------------
+
+/// The state of a request that is executed.
+pub struct Request<CR, Upstream, C>
+where
+    CR: Send + Sync,
+    Upstream: Send + Sync,
+    C: Clock + Send + Sync,
+{
+    /// State of the request.
+    state: RequestState,
+
+    /// The request message.
+    request_msg: CR,
+
+    /// The upstream transport of the connection.
+    upstream: Upstream,
+
+    /// The cache of the connection.
+    cache: Cache<Key, Arc<Value<C>>>,
+
+    /// The configuration of the connection.
+    config: Config,
+
+    /// The clock to use for expiring cache entries.
+    clock: C,
+}
+
+impl<CR, Upstream, C> Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Send + Sync,
+    Upstream: SendRequest<CR> + Send + Sync,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    /// Create a new Request object.
+    fn new(
+        request_msg: CR,
+        upstream: Upstream,
+        cache: Cache<Key, Arc<Value<C>>>,
+        config: Config,
+        clock: C,
+    ) -> Request<CR, Upstream, C> {
+        Self {
+            state: RequestState::Init,
+            request_msg,
+            upstream,
+            cache,
+            config,
+            clock,
+        }
+    }
+
+    /// This is the implementation of the get_response method.
+    // This function is cancel safe.
+    async fn get_response_impl(&mut self) -> Result<Message<Bytes>, Error> {
+        loop {
```

Comment:
Why does this loop?

Reply:
Otherwise the code gets too complex, in particular

self.state =
                            RequestState::GetResponseNoCache(request);
                        continue;

Reply:
To better illustrate my thinking see https://github.com/NLnetLabs/domain/pull/282.

Reply:
Ah, that PR misses that your code might use a state machine not for logic reasons but purely to save state across an .await point. If that's the case a comment to that effect somewhere would be very useful, otherwise it is unclear why the code is structured that way.

Reply:
Minor idea: If you switch the state to an option, you can switch the loop to a `while let Some` kind of thing and automatically fuse the future.